### PR TITLE
Pass `stimulus_id` to `SchedulerPlugin.remove_worker` and `SchedulerPlugin.transition`

### PR DESCRIPTION
**Pull request by hendrikmakait**
Addresses #7973 by implementing the bare minimum to support #7970 

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks @hendrikmakait. This will break existing plugins that have a `transition` and/or `remove_worker` method. Can we make `stimulus_id` optional such that existing plugins continue to work?

**Comment by hendrikmakait** [comment]
> Thanks @hendrikmakait. This will break existing plugins that have a `transition` and/or `remove_worker` method. Can we make `stimulus_id` optional such that existing plugins continue to work?

I'm not sure if we can do this _nicely_. The problem is that some plugins might accept it as a keyword, others might not, so (I think) we'd have to resort to introspection, which is rather cumbersome. https://github.com/dask/distributed/blob/8e92b8533a81cb52e2893f0b14c4e7842d2f607a/distributed/scheduler.py#L1981-L1993 implements this. I think I'd be fine with adding such a code for a deprecation cycle and issue a `DeprecationWarning` if `stimulus_id` is not supported, but I'd rather want to avoid this long-term.

**Comment by hendrikmakait** [comment]
@jrbourbeau: I've added a FutureWarning for `remove_worker`. Since `transition` supports `**kwargs`, there's no strict need, though I'm wondering if we should still add a warning to alert users that the content of `kwargs` might change? I might be overthinking this.

**Comment by fjetter** [comment]
The entire purpose of `kwargs` is to tell people to implement their plugins in a forward compatible way. As long as we specify kwargs in the signature, we can freely add new paramters without affecting old users. Given our fragile plugin-API surface I suggest to keep kwargs and don't worry about any warnings for methods that had this in before.

Alerting about the content of kwargs seems counterintutive to me. If somebody cares aobut this, they should specify it as a proper argument

File: distributed/diagnostics/plugin.py [hunk 1]
```diff
@@ -164,7 +167,7 @@ def add_worker(self, scheduler: Scheduler, worker: str) -> None | Awaitable[None
         """
 
     def remove_worker(
-        self, scheduler: Scheduler, worker: str
+        self, scheduler: Scheduler, worker: str, *, stimulus_id: str
     ) -> None | Awaitable[None]:
```
Comment (fjetter) [inline distributed/diagnostics/plugin.py:171]:
```suggestion
    def remove_worker(
        self, scheduler: Scheduler, worker: str, stimulus_id: str, **kwargs
    ) -> None | Awaitable[None]:
```

I suggest to not force the keyword and allow kwargs for future compat reasons, see https://github.com/dask/distributed/pull/7974#issuecomment-1632295809

File: docs/source/plugins.rst [hunk 2]
```diff
@@ -38,7 +38,7 @@ for more information on RabbitMQ and how to consume the messages.
            self.channel = self.connection.channel()
            self.channel.queue_declare(queue='dask_task_status', durable=True)
 
-       def transition(self, key, start, finish, *args, **kwargs):
+       def transition(self, key, start, finish, *args, stimulus_id, **kwargs):
```
Comment (fjetter) [inline docs/source/plugins.rst:41]: This is another great example why we should just require/encourage kwargs. I don't think this should be mandatory

**Comment by hendrikmakait** [comment]
> The entire purpose of `kwargs` is to tell people to implement their plugins in a forward compatible way. As long as we specify kwargs in the signature, we can freely add new paramters without affecting old users. Given our fragile plugin-API surface I suggest to keep kwargs and don't worry about any warnings for methods that had this in before.
> 
> Alerting about the content of kwargs seems counterintutive to me. If somebody cares aobut this, they should specify it as a proper argument

Makes sense, I wasn't sure if the purpose of `kwargs` was to encourage forwrad-compatible implementations or to mimic the scheduler API. In the case of `transition`, `Scheduler.transition` itself takes `kwargs` because different transitions take different inputs.

I'll adjust this accordingly.

File: distributed/diagnostics/plugin.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/diagnostics/plugin.py:171]: Forcing the keyword while adding `**kwargs` would make a method with the signature `MySchedulerPlugin.remove_worker(self, scheduler, worker, **kwargs)` fully forward-compatible, so I'd suggest sticking with it unless we also want to require `*args` in the signature.

File: distributed/scheduler.py [hunk 3]
```diff
@@ -5725,6 +5736,15 @@ def add_plugin(
                 category=UserWarning,
             )
 
+        parameters = inspect.signature(plugin.remove_worker).parameters
+        if "kwargs" not in parameters:
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5740]: Support ``**kwds``. warn on ``kwargs`` without ``**``.

**Review by crusaderky — CHANGES_REQUESTED** [review]
Minor tweak on deprecation

File: distributed/scheduler.py [hunk 4]
```diff
@@ -5070,7 +5072,16 @@ async def remove_worker(
...
+                try:
+                    result = plugin.remove_worker(
+                        scheduler=self, worker=address, stimulus_id=stimulus_id
+                    )
+                except TypeError as e:
+                    parameters = inspect.signature(plugin.remove_worker).parameters
+                    if "stimulus_id" not in parameters and "kwargs" not in parameters:
+                        result = plugin.remove_worker(scheduler=self, worker=address)  # type: ignore
+                    else:
+                        raise e
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:5084]: Good point, thanks!

**Comment by hendrikmakait** [comment]
@crusaderky: All feedback has been addressed and a test added for `**kwargs` renaming. This is ready for another review.

**Review by fjetter — APPROVED** [review]

**Pull request merged**